kinoml/core/protein.py

```diff
@@ -0,0 +1,350 @@
+    @classmethod
+    def _from_id_klifs(cls, identifier, **kwargs):
+        raise NotImplementedError
+
+    @classmethod
+    def _from_id_kinmap(cls, identifier, **kwargs):
+        raise NotImplementedError
+
+    @classmethod
+    def _from_id_manning(cls, identifier, **kwargs):
+        raise NotImplementedError
+
+
+class Biosequence(str):
+    """
+    Base class for string representations of biological polymers
+    (nucleic acids, peptides, proteins...)
+    TODO: How to handle several mutations at the same time, while
+          keeping indices relevant (after a deletion, a replacement
+          or insertion position might be wrong).
+    """
+
+    ALPHABET = set(ascii_letters)
+    _ACCESSION_URL = None
+    ACCESSION_MAX_RETRIEVAL = 50
+
+    def __new__(cls, value, header="", _provenance=None, *args, **kwargs):
+        diff = set(value).difference(cls.ALPHABET)
+        if diff:
+            raise ValueError(
+                f"Biosequence can only contain characters in {cls.ALPHABET}, "
+                f"but found these extra ones: {diff}."
+            )
+        s = super().__new__(cls, value, *args, **kwargs)
+        s.header = header
+        s._provenance = {}
+        # TODO: We might override some provenance data with this blind update
+        if _provenance is not None:
+            s._provenance.update(_provenance)
+        return s
+
+    @classmethod
+    def from_accession(cls, *accessions):
+        """
+        Get FASTA sequence from an online accession identifier
+
+        Parameters
+        ----------
+        accession : str
+            NCBI identifier. Multiple can be provided!
+        Returns
+        -------
+        Biosequence or list of Biosequencs
+        """
+        if cls._ACCESSION_URL is None:
+            raise NotImplementedError
+        if len(accessions) > cls.ACCESSION_MAX_RETRIEVAL:
+            raise ValueError(
+                f"You can only provide {cls.ACCESSION_MAX_RETRIEVAL} accessions at the same time."
+            )
+        r = requests.get(cls._ACCESSION_URL.format(",".join(accessions)))
+        r.raise_for_status()
+        sequences = []
+        for line in r.text.splitlines():
+            line = line.strip()
+            if not line:
+                continue
+            if line.startswith(">"):
+                sequences.append({"header": line[1:], "sequence": []})
+            else:
+                sequences[-1]["sequence"].append(line)
+        if not sequences:
+            return
+        objects = []
+        for sequence, accession in zip(sequences, accessions):
+            obj = cls(
+                "".join(sequence["sequence"]),
+                header=sequence["header"],
+                _provenance={"accession": accession},
+            )
+            objects.append(obj)
+        if not objects:
+            return None
+        if len(objects) == 1:
+            return objects[0]
+        return objects
+
+    def cut(self, start, stop, check=True):
```

Comment:
Protein constructs are not generally referred to via notation like `C123`, `T234`, but instead giving the amino acid numerical span for a given reference sequence. Maybe we could use a more natural way to do this?

Reply:
Sure, I have no strong feelings about this. This was built to support the kind of data we have for Kinomescan, but if it makes more sense to use support another syntax, I am all in.